File: src/data/courses/dsa/string-hashing-problems/hashing-based-problems/longest-consecutive-sequence/code/python/longest_consecutive_sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # O(n) time, O(n) space — HashSet, start from sequence beginnings
        num_set = set(nums)
        max_len = 0
        
        for num in num_set:
            # Only start counting from sequence beginning
            if num - 1 not in num_set:
                cur, length = num, 1
                while cur + 1 in num_set:
                    cur += 1
                    length += 1
                max_len = max(max_len, length)
        
        return max_len

File: src/data/courses/dsa/string-hashing-problems/hashing-based-problems/longest-consecutive-sequence/code/cpp/longest_consecutive_sequence.cpp
class Solution {
public:
    // O(n) time, O(n) space — HashSet + only start from sequence beginnings
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> numSet(nums.begin(), nums.end());
        int maxLen = 0;
        
        for (int num : numSet) {
            // Only start counting if num is the START of a sequence
            // (i.e., num-1 is NOT in the set)
            if (!numSet.count(num - 1)) {
                int currentNum = num;
                int currentLen = 1;
                
                // Count consecutive numbers
                while (numSet.count(currentNum + 1)) {
                    currentNum++;
                    currentLen++;
                }
                maxLen = max(maxLen, currentLen);
            }
        }
        return maxLen;
    }
};